### ISSUE TITLE
BBOX_STD_DEV usage is unclear and possibly incorrect?

### ISSUE THREAD

**Issue by simonm3** [matterport/Mask_RCNN#270]
This is really great project - love the level of comments, the working notebooks, and the simple shapes dataset.....all makes it exceptionally easy to read.

One thing that is not clear - In the RPN proposal layer the deltas are multiplied by BBOX_STD_DEV. There was a previous issue asking about this https://github.com/matterport/Mask_RCNN/issues/85.  The answer given refers to fastrcnn paper section on normalising the regression targets for the loss function.

However this is applied after the RPN loss function; and before the bbox spec is normalised to 0-1 range. Also two of the delta numbers are log(delta). I can possibly envisage that you might divide by the standard deviation as part of normalisation. However this is multiplying; by a log; and not in the right place for that.

**Comment by FruVirus** [comment]
Do you know where the exact values for BBOX_STD_DEV comes from?

**Comment by FruVirus** [comment]
Looking at the threads on the Faster R-CNN channels, it seems this question was never answered either. I suppose one way to test it out is just to train on a dataset and try it with and without the standard deviation normalization and play around with the numbers.

I get the feeling that these numbers were empirically found by R. Girshick when he was training on the COCO dataset and is likely to be different depending on the particular dataset.

**Comment by waleedka** [comment]
@simonm3 The regressor performs best if its outputs have a mean of zero and a standard deviation of one. To achieve that we need to do two transformations:

1. When preparing the training targets for the regressor, subtract the mean and divide by the standard deviation. This will give the regressor normalized training targets. That also means that the regressor is going to predict normalized values rather than the actual values we would use to shift and resize the anchors.
2. After training is complete, when we want to use the output from the regressor, we multiply by the standard deviation and add the mean to reverse the transformation in 1. This converts the predicted normalized values into real values we can use to shift and resize the anchor boxes.

Knowing that anchors cover the image uniformly, it's fair to assume that shifts and resizes are evenly distributed between positive and negative, so we assume the mean is close to zero already, and skip the part I mentioned above about subtracting the mean.

For the standard deviation, ideally we'd measure it across the dataset. In this case, as @FruVirus guessed, I used the same values that R. Girshick used in Faster RCNN. From [their config file](https://github.com/rbgirshick/py-faster-rcnn/blob/96dc9f1dea3087474d6da5a98879072901ee9bf9/lib/fast_rcnn/config.py):
```
# Normalize the targets using "precomputed" (or made up) means and stdevs
# (BBOX_NORMALIZE_TARGETS must also be True)
__C.TRAIN.BBOX_NORMALIZE_TARGETS_PRECOMPUTED = False
__C.TRAIN.BBOX_NORMALIZE_MEANS = (0.0, 0.0, 0.0, 0.0)
__C.TRAIN.BBOX_NORMALIZE_STDS = (0.1, 0.1, 0.2, 0.2)
```

**Comment by simonm3** [comment]
Thank for great explanation.

**Comment by CMCDragonkai** [comment]
How would you measure the standard-deviation if you really wanted to?

**Comment by patrick-12sigma** [comment]
@CMCDragonkai I suppose you can set the std values in `BBOX_STD_DEV` all to 1, and then compute the stdev of the relative shift. Concretely, in the data generator section of `inspect_data.ipynb`, after you get the `rpn_bbox`,

```python
std_list = []
for i in range(4):
    a = rpn_bbox[:, :, i]
    std_list.append(np.std(a[np.nonzero(a)]))
print(std_list)
```
gives the std list.

**Comment by CMCDragonkai** [comment]
But there are 2:

```
    # Bounding box refinement standard deviation for RPN and final detections.
    RPN_BBOX_STD_DEV = np.array([0.1, 0.1, 0.2, 0.2])
    BBOX_STD_DEV = np.array([0.1, 0.1, 0.2, 0.2])
```

If we measure it exactly as in your example, are you saying we should use those as our hyperparameters?

**Comment by gaklein2007** [comment]
The anchors are uniform across the image, but what are being normalized are the deltas to transform the anchors to the correct bounding box. That being said, why would the deltas have a mean of zero? This is only the case if a bounding box is located in the center. If the point of interest if located off center, then wouldn't you also need to subtract the mean deltas?

**Comment by patrick-12sigma** [comment]
@gaklein2007 the deltas are the difference in coordinates to the nearest anchor box (or the anchor box with the best IoU with the bbox in question), not only the anchor box in the center. The relative shift of any bbox to the nearest anchor box is uniformly random, therefore the mean is roughly zero.

**Comment by gaklein2007** [comment]
@patrick-12sigma I am not referring to a anchor box in the center, but a ground truth bounding box. I understand the distance between any bounding box and its closest neighbors will be an average of zero as there would be an equal amount surrounding a bounding box from all sides. However, when making the rpn deltas in the generator, negative anchors are also added to add background into it and since the bounding box target is not necessarily center, these do not necessarily have to average out to zero, especially if going from sample to sample the object of interest changes position.

**Comment by gaklein2007** [comment]
Actually, disregard that last point as I forgot for the deltas negatives are not considered, only positive anchors.

**Comment by patrick-12sigma** [comment]
That is right. Deltas for negative cases are not accounted for in the regression loss.

**Comment by destinyzs** [comment]
If my regression is `[y, x, h, w, theta]`, like something rotate.
How should i set the rotate `BBOX_NORMALIZE_MEANS` and `BBOX_NORMALIZE_STDS` ?
Still `[0.0, 0.0, 0.0, 0.0, 0.0]` and  `[0.1, 0.1, 0.2, 0.2, 0.2]`?

**Comment by patrick-12sigma** [comment]
@gaklein2007 for very small bboxes you should change your anchor box scale to match the bboxes you would like to detect. I would do a histogram of the groundtruth bbox sizes to make sure the anchor scales have a good match with that.

**Comment by gaklein2007** [comment]
I have been tweaking the scales but without much luck. I also changed a part of the code that had a TODO on it, which actually made it significantly worse. After changing it back to the original version and just padding the coordinates of the bounding boxes, I am testing it to see if that was why the FRCNN (not attempting segmentations yet) was failing. I believe that was the cause of the data issues.

**Comment by patrick-12sigma** [comment]
In general I would first give a few examples for the network to learn. We expect the network would memorize the examples, and if you evaluate the performance on the training set, you should get almost perfect score. Otherwise it means the pipeline is broken, for example the data loading part.

**Comment by gaklein2007** [comment]
You mean pretraining it on some samples that have an iou > 0.7 and then feeding in the rest of the data?

**Comment by patrick-12sigma** [comment]
@gaklein2007  I am talking about feeding one or two images to the network first to make sure the network is able to overfit to your limited training examples.

**Comment by leimao** [comment]
> @simonm3 The regressor performs best if its outputs have a mean of zero and a standard deviation of one. To achieve that we need to do two transformations:
> 
> 1. When preparing the training targets for the regressor, subtract the mean and divide by the standard deviation. This will give the regressor normalized training targets. That also means that the regressor is going to predict normalized values rather than the actual values we would use to shift and resize the anchors.
> 2. After training is complete, when we want to use the output from the regressor, we multiply by the standard deviation and add the mean to reverse the transformation in 1. This converts the predicted normalized values into real values we can use to shift and resize the anchor boxes.
> 
> Knowing that anchors cover the image uniformly, it's fair to assume that shifts and resizes are evenly distributed between positive and negative, so we assume the mean is close to zero already, and skip the part I mentioned above about subtracting the mean.
> 
> For the standard deviation, ideally we'd measure it across the dataset. In this case, as @FruVirus guessed, I used the same values that R. Girshick used in Faster RCNN. From [their config file](https://github.com/rbgirshick/py-faster-rcnn/blob/96dc9f1dea3087474d6da5a98879072901ee9bf9/lib/fast_rcnn/config.py):
> 
> ```
> # Normalize the targets using "precomputed" (or made up) means and stdevs
> # (BBOX_NORMALIZE_TARGETS must also be True)
> __C.TRAIN.BBOX_NORMALIZE_TARGETS_PRECOMPUTED = False
> __C.TRAIN.BBOX_NORMALIZE_MEANS = (0.0, 0.0, 0.0, 0.0)
> __C.TRAIN.BBOX_NORMALIZE_STDS = (0.1, 0.1, 0.2, 0.2)
> ```

waleedka's explanation is helpful.
